### PR TITLE
Feature/sg 000 qat

### PR DESCRIPTION
**Pull request by NatanBagrov**

### CODE REVIEW

File: src/super_gradients/training/utils/quantization/core.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+from dataclasses import dataclass
+from enum import Enum
+from typing import Union, Type, Optional
+
+from pytorch_quantization.nn.modules._utils import QuantMixin, QuantInputMixin
+from pytorch_quantization.tensor_quant import QuantDescriptor
+from torch import nn
+
+
+def _extract_init_args(cls, float_instance):
```
Comment (shaydeci) [inline src/super_gradients/training/utils/quantization/core.py:11]: Are we assuming here that the arguement keys must match the attribute names of the module?

If so, this is can be problematic...In that case we should consider allowing different _extract_init_args that are block specific.

File: src/super_gradients/training/utils/quantization/core.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    if 'bias' in float_instance_state:
+        if float_instance_state['bias'] is None:  # None is the state when bias=False in torch.nn
+            float_instance_state['bias'] = False
+        elif not isinstance(float_instance_state['bias'], bool):  # Tensor is the state when bias=True in torch.nn
+            float_instance_state['bias'] = True
+        # in case bias is a boolean - we don't do anything, so it is taken as-is, either True or False
+    return float_instance_state
+
+
+def _from_float(cls, float_instance, **kwargs):
```
Comment (shaydeci) [inline src/super_gradients/training/utils/quantization/core.py:39]: If my question in my first comment is true, then we need to allow passing _extract_init_args to this function as well.

File: src/super_gradients/training/utils/quantization/core.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    def __init__(self, *, float_module: nn.Module,
+                 quantized_type: Union[Type[QuantMixin], Type[QuantInputMixin], Type[SGQuantMixin]],
+                 input_quant_descriptor: QuantDescriptor = None,
+                 weights_quant_descriptor: QuantDescriptor = None) -> None:
+        super().__init__()
+        self.float_module = float_module
+        self.quantized_type = quantized_type
+        self.input_quant_descriptor = input_quant_descriptor,
+        self.weights_quant_descriptor = weights_quant_descriptor,
+        self.forward = float_module.forward
```
Comment (shaydeci) [inline src/super_gradients/training/utils/quantization/core.py:95]: Can we rewrite this inside this modules forward? It's a bit hacky..

File: src/super_gradients/training/utils/quantization/fine_grain_quantization_utils.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        if _imported_pytorch_quantization_failure is not None:
+            raise _imported_pytorch_quantization_failure
+        for candidate_key in ('.'.join(nesting + (child_name,)), type(child_module)):
+            if candidate_key not in mapping_instructions:
+                continue
+            metadata: QuantizedMetadata = mapping_instructions[candidate_key]
+            if metadata.action == QuantizedMetadata.ReplacementAction.SKIP:
+                return True
+            elif metadata.action == QuantizedMetadata.ReplacementAction.UNWRAP:
+                assert isinstance(child_module, SkipQuantization)
```
Comment (shaydeci) [inline src/super_gradients/training/utils/quantization/fine_grain_quantization_utils.py:167]: Why the assert ?

File: tests/unit_tests/quantization_utility_tests.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    from super_gradients.training.utils.quantization.core import SkipQuantization, SGQuantMixin, QuantizedMapping
+
+    _imported_pytorch_quantization_failure = None
+
+except (ImportError, NameError, ModuleNotFoundError) as import_err:
+    _imported_pytorch_quantization_failure = import_err
+
+
+@unittest.skipIf(_imported_pytorch_quantization_failure is not None, "Failed to import `pytorch_quantization`")
+class QuantizationUtilityTest(unittest.TestCase):
```
Comment (shaydeci) [inline tests/unit_tests/quantization_utility_tests.py:18]: We need to add some example code for usage. Can't assume users will search the tests directory for usage example.

**Review by shaydeci — CHANGES_REQUESTED** [review]
First review, I guess many changes are coming.
Yet besides the inline comments:
First and foremost- we can't assume our users will go through the tests directory for usage:
- Make sure to add some example code for usage at the module level.
- Make sure to add some example code for the usage in SgModel level (i.e following the same flow in our current QAT example).

Note that its important that we allow keep the layers visible (i.e if user print the network they should not see some weird named blocks). I know that with the decorators we preserve the visibility, so perhaps use them when writing the examples.

On a different note: make sure to add what the PR includes (very important for the reviewer when you split PRs), and what comes next.

File: src/super_gradients/training/utils/quantization/core.py (hunk 1 above)
Comment (NatanBagrov) [reply src/super_gradients/training/utils/quantization/core.py:11]: We do make that assumption.
This functionality is the default and is used if `SGQuantMixin.from_float` is not overridden.
To do custom stuff, one may override `SGQuantMixin.from_float`

File: src/super_gradients/training/utils/quantization/core.py (hunk 3 above)
Comment (NatanBagrov) [reply src/super_gradients/training/utils/quantization/core.py:95]: Sure we can, however, I want to be invariant for those weird `forward`s that accept more than one argument.
We can use `*args` though... whatever you like

File: src/super_gradients/training/utils/quantization/fine_grain_quantization_utils.py (hunk 4 above)
Comment (NatanBagrov) [reply src/super_gradients/training/utils/quantization/fine_grain_quantization_utils.py:167]: Just a sanity that we don't unwrap non-wrapper modules

File: src/super_gradients/training/utils/quantization/core.py (hunk 2 above)
Comment (NatanBagrov) [reply src/super_gradients/training/utils/quantization/core.py:39]: To do custom stuff, one may override `SGQuantMixin.from_float`

File: tests/unit_tests/quantization_utility_tests.py (hunk 5 above)
Comment (NatanBagrov) [reply tests/unit_tests/quantization_utility_tests.py:20]: Added examples in [3c93d62](https://github.com/Deci-AI/super-gradients/pull/304/commits/3c93d622af99ac6d68ba532b5c8f939ef67ef20e)

File: src/super_gradients/examples/resnet_qat/resnet_qat_example.py [hunk 6]
```diff
@@ -20,40 +20,41 @@
 from super_gradients.training.metrics.classification_metrics import Accuracy
 
 import super_gradients
-from super_gradients.training.utils.quantization_utils import PostQATConversionCallback
+from super_gradients.training.utils.quantization_callbacks import PostQATConversionCallback
```
Comment (shaydeci) [inline src/super_gradients/examples/resnet_qat/resnet_qat_example.py:23]: Remove as it isn't supported at the moment

File: src/super_gradients/examples/quantization/resnet_quantization_examples.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+from super_gradients.training.utils.quantization.calibrator import QuantizationCalibrator
+from super_gradients.training.utils.quantization.core import SGQuantMixin, QuantizedMetadata
+from super_gradients.training.utils.quantization.export import export_quantized_module_to_onnx
+from super_gradients.training.utils.quantization.selective_quantization_utils import SelectiveQuantizer
+
+
+def _calibrate_and_export(module, method, onnx_name):
+    module = module.cuda()
+    # CALIBRATE (PTQ)
+    dataset_interface = ImageNetDatasetInterface(
```
Comment (shaydeci) [inline src/super_gradients/examples/quantization/resnet_quantization_examples.py:20]: Old- shouldn't work...remove

File: src/super_gradients/examples/quantization/resnet_quantization_examples.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
+from collections import OrderedDict
```
Comment (shaydeci) [inline src/super_gradients/examples/quantization/resnet_quantization_examples.py:1]: Remove/remake the entire file please. Many irrelevant old parts which would crash.

File: src/super_gradients/training/utils/quantization/export.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    model.eval()
+    if hasattr(model, "prep_model_for_conversion"):
+        model.prep_model_for_conversion()
+    quant_nn.TensorQuantizer.use_fb_fake_quant = True
+    # Export ONNX for multiple batch sizes
+    logger.info("Creating ONNX file: " + onnx_filename)
+    dummy_input = torch.randn(input_shape, device=next(model.parameters()).device)
+    torch.onnx.export(model, dummy_input, onnx_filename, verbose=False, opset_version=13, enable_onnx_checker=False, do_constant_folding=True)
+    # Restore functions of quant_nn back as expected
+    quant_nn.TensorQuantizer.use_fb_fake_quant = False
```
Comment (shaydeci) [inline src/super_gradients/training/utils/quantization/export.py:36]: Save the  quant_nn.TensorQuantizer.use_fb_fake_quant in the start of the ruction and modify it back to its initial value (of calling this method).

File: src/super_gradients/examples/quantization/selective_quantization_examples.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
+import torch
+from torch import nn
+
+from super_gradients.training.utils.quantization.calibrator import QuantizationCalibrator
+from super_gradients.training.datasets import Cifar10DatasetInterface
```
Comment (shaydeci) [inline src/super_gradients/examples/quantization/selective_quantization_examples.py:5]: Remove datasetinterface from here and any usages of it.

File: src/super_gradients/examples/quantization/selective_quantization_examples.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    with torch.no_grad():
+        y = module(x)
+        torch.testing.assert_close(y.size(), (1, n_clss))
+
+    # EXPORT TO ONNX
+    export_quantized_module_to_onnx(module, "my_quantized_model.onnx", input_shape=(1, 3, res, res))
+
+
+if __name__ == "__main__":
+    vanilla_quantize_all_example()
```
Comment (shaydeci) [inline src/super_gradients/examples/quantization/selective_quantization_examples.py:184]: Split the different examples into different pages please.

File: src/super_gradients/examples/quantization/selective_quantization_examples.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    non_default_calibrators_example()
```
Comment (shaydeci) [inline src/super_gradients/examples/quantization/selective_quantization_examples.py:185]: Also add example using decorators.

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 13]
```diff
@@ -1047,6 +1054,9 @@ def forward(self, inputs, targets):
         # ADD CALLBACK FOR QAT
         self.enable_qat = core_utils.get_param(self.training_params, "enable_qat", False)
         if self.enable_qat:
+            if _imported_quantization_failure is not None:
+                raise ImportError(_imported_quantization_failure)
```
Comment (shaydeci) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:1058]: Raise error for enable QAT.

File: tests/deci_core_unit_test_suite_runner.py [hunk 14]
```diff
@@ -85,6 +86,7 @@ def _add_modules_to_unit_tests_suite(self):
         self.unit_tests_suite.addTest(self.test_loader.loadTestsFromModule(IoULossTest))
         self.unit_tests_suite.addTest(self.test_loader.loadTestsFromModule(TestDetectionDatasetSubsampling))
         self.unit_tests_suite.addTest(self.test_loader.loadTestsFromModule(TestDetectionDatasetSubclassing))
+        self.unit_tests_suite.addTest(self.test_loader.loadTestsFromModule(QuantizationUtilityTest))
```
Comment (shaydeci) [inline tests/deci_core_unit_test_suite_runner.py:89]: Lets check the dependencies for pytorch-quantization + licenses, if there are no req for GPUs, then lets add it to our requirements and run the unit tests in the CI.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Inline, we're almost there.

File: src/super_gradients/training/utils/quantization/quantized_modules/pytorch_quantization_resnet.py [hunk 15]
```diff
@@ -0,0 +1,404 @@
+"""
+Quantized ResNet-50, based on pytorch-quantization TensorRT tools
+Based on:
+https://github.com/NVIDIA/TensorRT/blob/
+51a4297753d3e12d0eed864be52400f429a6a94c/tools/pytorch-quantization/examples/torchvision/models/classification/resnet.py
+(Licensed under the Apache License, Version 2.0)
+"""
```
[the remark this answers is not in the document]
Comment (ofrimasad) [reply src/super_gradients/training/utils/quantization/quantized_modules/pytorch_quantization_resnet.py:7]: please drop this file

**Review by ofrimasad — COMMENTED** [review]
great work! 
a few comments

File: src/super_gradients/training/utils/quantization/core.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+    def __init__(
+        self,
+        *,
+        float_module: nn.Module,
+        quantized_target_class: Union[Type[QuantMixin], Type[QuantInputMixin], Type[SGQuantMixin]],
+        action=QuantizedMetadata.ReplacementAction.REPLACE_AND_DONT_QUANTIZE_CHILD_MODULES,
+        input_quant_descriptor: QuantDescriptor = None,
+        weights_quant_descriptor: QuantDescriptor = None,
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (spsancti) [reply src/super_gradients/training/utils/quantization/core.py:159]: Let's leave it for now, as they have different functions and change to merge them is too big and affects almost everything

File: src/super_gradients/training/utils/quantization/selective_quantization_utils.py [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+
+    def register_quantization_mapping(
+        self, *, layer_names: Set[str], quantized_target_class: Type[SGQuantMixin], input_quant_descriptor=None, weights_quant_descriptor=None
+    ):
+        self.mapping_instructions.update(
+            {
+                name: QuantizedMetadata(
+                    float_source=name,
+                    quantized_target_class=quantized_target_class,
+                    action=QuantizedMetadata.ReplacementAction.REPLACE_AND_DONT_QUANTIZE_CHILD_MODULES,
```
[the remark this answers is not in the document]
Comment (spsancti) [reply src/super_gradients/training/utils/quantization/selective_quantization_utils.py]: Let's leave it as is and refactor along with QuantizedMetadata

File: requirements.txt [hunk 18]
```diff
@@ -32,4 +31,5 @@ packaging>=20.4
 wheel>=0.38.0
 # not directly required, pinned by Snyk to avoid a vulnerability
 pygments>=2.7.4
+--extra-index-url https://pypi.ngc.nvidia.com
```
Comment (spsancti) [inline requirements.txt:34]: This one still fails in circleci because it reads requirements line-by-line to get rid of the comments
We have 
pip install --extra-index-url https://pypi.ngc.nvidia.com
pip install pytorch-quantization==2.1.2

Comment (spsancti) [reply requirements.txt:34]: Why do we need comments in requirements.txt?

Comment (spsancti) [reply requirements.txt:34]: Why do we need a special way of reading requirements.txt if comments are the part of PEP?

**Review by ofrimasad — COMMENTED** [review]
looks very nice!
one minor questions

**Review by shaydeci — APPROVED** [review]
